indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,685 @@
+      if (tryInterval.contains(request.getRowInterval())) {
+        tryIntervalToRequests.computeIfAbsent(tryInterval, i -> new ArrayList<>()).add(request);
+      }
+    }
+    return tryIntervalToRequests;
+  }
+
+  private void emitTaskMetric(String metric, long value, SegmentAllocateRequest request)
+  {
+    final ServiceMetricEvent.Builder metricBuilder = ServiceMetricEvent.builder();
+    IndexTaskUtils.setTaskDimensions(metricBuilder, request.getTask());
+    metricBuilder.setDimension("taskActionType", SegmentAllocateAction.TYPE);
+    emitter.emit(metricBuilder.build(metric, value));
+  }
+
+  private void emitBatchMetric(String metric, long value, AllocateRequestKey key)
+  {
+    final ServiceMetricEvent.Builder metricBuilder = ServiceMetricEvent.builder();
+    metricBuilder.setDimension("taskActionType", SegmentAllocateAction.TYPE);
+    metricBuilder.setDimension(DruidMetrics.DATASOURCE, key.dataSource);
+    metricBuilder.setDimension(DruidMetrics.INTERVAL, key.preferredAllocationInterval.toString());
+    emitter.emit(metricBuilder.build(metric, value));
+  }
+
+  public void becomeLeader()
+  {
+    if (isEnabled()) {
+      // Start polling the queue
+      log.info("Elected leader. Starting queue processing.");
+      scheduleQueuePoll(maxWaitTimeMillis);
+    } else {
+      log.info("Elected leader but batched segment allocation is disabled.");
+    }
+  }
+
+  public void stopBeingLeader()
+  {
+    log.info("Not leader anymore. Stopping queue processing.");
+  }
+
+  /**
+   * A batch of segment allocation requests.
+   */
+  private class AllocateRequestBatch
+  {
+    private long queueTimeMillis;
+    private final AllocateRequestKey key;
+
+    /**
+     * This must be accessed through methods synchronized on this batch.
+     * It is to avoid races between a new request being added just when the batch
+     * is being processed.
+     */
+    private final Map<SegmentAllocateRequest, CompletableFuture<SegmentIdWithShardSpec>>
+        requestToFuture = new HashMap<>();
+
+    AllocateRequestBatch(AllocateRequestKey key)
+    {
+      log.info("Creating a new batch with key: %s", key);
+      this.key = key;
+    }
+
+    synchronized Future<SegmentIdWithShardSpec> add(SegmentAllocateRequest request)
+    {
+      log.info("Adding request to batch [%s]: %s", key, request.getAction());
+      return requestToFuture.computeIfAbsent(request, req -> new CompletableFuture<>());
+    }
+
+    synchronized void merge(AllocateRequestBatch batch)
+    {
+      requestToFuture.putAll(batch.requestToFuture);
+      batch.requestToFuture.clear();
+    }
+
+    synchronized List<SegmentAllocateRequest> getRequests()
+    {
+      return new ArrayList<>(requestToFuture.keySet());
+    }
+
+    synchronized void markCompleted()
+    {
+      if (!requestToFuture.isEmpty()) {
+        log.info("Marking [%d] requests in batch [%s] as failed.", size(), key);
+        requestToFuture.values().forEach(future -> future.complete(null));
```

Comment:
shouldn't this be future.completeExceptionally?

Reply:
invoking future calls can also result in an exception. Ideally, the 4 lines below should be done first. Or you can wrap the `future.complete` in a try/catch block.

Reply:
> shouldn't this be future.completeExceptionally?

No, this is in order to retain the existing behaviour.

Reply:
> invoking future calls can also result in an exception. Ideally, the 4 lines below should be done first. Or you can wrap the future.complete in a try/catch block.

Oh, I didn't realize. Are you sure? I am just completing the future here and setting a new value in it. 
The only buggy scenario I can think is trying to complete an already completed future, but it doesn't throw an exception even then.

Reply:
I meant that there can be an exception in a listener attached to this future. e.g. 
```
future.whenComplete() {
  throw some exception; 
}
```

But if the callback is a simple op, its fine.